include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,692 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__quadrature_point_data_h
+#define dealii__quadrature_point_data_h
+
+#ifdef DEAL_II_WITH_CXX11
+
+#include <deal.II/base/quadrature.h>
+#include <deal.II/base/config.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/fe/fe_tools.h>
+#include <deal.II/fe/fe.h>
+#include <deal.II/lac/vector.h>
+
+#include <vector>
+#include <map>
+#include <typeinfo>
+
+DEAL_II_NAMESPACE_OPEN
+
+/*!@addtogroup Quadrature */
+/*@{*/
+
+/**
+ * A class for storing at each cell @p CellIteratorType a vector of data @p DataType .
+ * The underlying structure and the initialize() method of this class is designed
+ * in such a way that one could use different child class derived from the
+ * base DataType class to store data on a given cell. This implies the usage of pointers, in our
+ * case -- std::shared_ptr().
+ *
+ * @note The limitation is that the data stored on each cell should be of the same type.
+ * For this reason, this class may not be sufficiently
+ * flexible when, for example, adopting a level-set approach to describe material behavior.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+template <typename CellIteratorType, typename DataType>
+class CellDataStorage
+{
+public:
+  /**
+   * A typedef for a vector of shared pointers with arbitrary data type @p T .
+   */
+  template <typename T> using CellVector = std::vector<std::shared_ptr<T> >;
+
+  /**
+   * Constructor.
+   */
+  CellDataStorage();
+
+  /**
+   * Destructor.
+   */
+  virtual ~CellDataStorage();
+
+  /**
+   * Initialize data on the @p cell to store @p n objects of type @p T .
+   * The typename @p T is possibly another class which is derived from the
+   * base @p DataType class. In order to initialize the vector of objects
+   * we have to assume that the class @p T has a standard constructor.
+   */
+  template<typename T=DataType>
+  void initialize(const CellIteratorType &cell,
+                  const unsigned int n);
```

Comment:
one could add `initialize(start, end, number)` to do `for (;start; start!=end; start++)` but then it would be a user responsibility to provide `start` and `end` such that we actually hit `end` at some point. Is it what you meant, @tjhei ?  Keep in mind that in future application one could have different data at different cells (e.g. plasticity + elasticity in material models), so one has to have a single-cell `initialize`.